packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+            } else if (action == 12) { // VAULT_DEPOSIT
+                address depositer; address vault; UFixed18 amount;
+                (depositer, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultDeposit(depositer, IPerennialVault(vault), amount);
+            } else if (action == 13) { // VAULT_REDEEM
+                address vault; UFixed18 shares;
+                (vault, shares, ptr) = decodeAddressAmount(input, ptr);
+
+                IPerennialVault(vault).redeem(shares, msg.sender);
+            } else if (action == 14) { // VAULT_CLAIM
+                address owner; address vault;
+                (owner, vault, ptr) = decodeAddressAddress(input, ptr);
+
+                IPerennialVault(vault).claim(owner);
+            } else if (action == 15) { // VAULT_WRAP_AND_DEPOSIT 
+                address account; address vault; UFixed18 amount;
+                (account, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultWrapAndDeposit(account, IPerennialVault(vault), amount);
+            }
+        }
+    }
+
+    /// Example Calldata Structure
+    /// let ptr
+    /// [ptr(userLen), ptr+1:userLen(user registry # OR 20 byte address if userLen == 0)] => address user 
+    /// ptr += (userLen OR 20) + 1
+    /// [ptr(prodcutLen), ptr+1:productLen(product registry # OR 20 byte address if prdoctLen == 0)] => address product
+    /// ptr += (prodcutLen OR 20) + 1
+    /// [ptr(amountLen), ptr:amountLen] => uint256 amount 
+
+
+    // ARGUMENT TYPE DECODING //
+
+    function decodeAddressAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, address addr2, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return (addr1, addr2, amount, ptr);
+    }
+
+    function decodeAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return(addr1, UFixed18(amount), ptr);
+    }
+
+    function decodeProductPrograms(bytes calldata input, uint ptr) private returns(address product, uint256[] memory programs, uint) {
+        (product, ptr) = decodeAddress(input, ptr);
+        (programs, ptr) = decodeUintArray(input, ptr);
+
+        return(product, programs, ptr);
+    }
+
+    function decodeAddressAddress(bytes calldata input, uint ptr) private returns(address addr1, address addr2, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+
+        return (addr1, addr2, ptr);
+    }
+
+
+    // INDIVIDUAL TYPE DECODING //
+
+     function decodeAmountUFixed18(bytes calldata input, uint ptr) private returns (UFixed18 result, uint) {
```

Comment:
nit: extra space before function

Reply:
once full comments added its nice to have section header an extra line above the first function

Reply:
Yeah, we used to do this kind of thing, but have standardized a bit per this style guideline: https://docs.soliditylang.org/en/develop/style-guide.html

Basically instead of having logical sections you use natspec + order by visibility and mutability (and use `_` prefix). This is feedback from auditors on what they prefer to look at.

Reply:
issue is that its all same visibility / mutability. i think its rlly nice to break argument type decoding and individual type decoding sections bc of this."

this the last thing before final review + deciding to rm readme (could add to dev docs if its rlly pertinent)

Reply:
i guess diff mutability for the address handling bc of cache but thats it

Reply:
ok lastly, in this vein do we redundantly comment for params for every function?